src/util/futurepool.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell::{Cell, RefCell, RefMut};
+use std::sync::{mpsc, Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::thread;
+use std::time::Duration;
+use futures::Future;
+use futures_cpupool::{self as cpupool, CpuFuture, CpuPool};
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send {
+    /// Will be invoked periodically (no less than specified interval).
+    /// When there is no task, it will NOT be invoked.
+    fn on_tick(&mut self) {}
+}
+
+/// A delegator to wrap `Context` to provide `on_tick` feature.
+#[derive(Debug)]
+struct ContextDelegator<T: Context> {
+    tick_interval: Duration,
+    inner: RefCell<T>,
+    last_tick: Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextDelegator<T> {
+    fn new(context: T, tick_interval: Duration) -> ContextDelegator<T> {
+        ContextDelegator {
+            tick_interval,
+            inner: RefCell::new(context),
+            last_tick: Cell::new(None),
+        }
+    }
+
+    fn get_context(&self) -> RefMut<T> {
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.get_context().on_tick();
+    }
+}
+
+/// Each `ContextDelegator` instance is invoked individually for each thread.
+/// It is never accessed concurrently so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextDelegator<T> {}
```

Comment:
I still don't get it why it should be `Sync`. If it will never be accessed concurrently, then you should mark the user as `Sync`.

Reply:
Please check out this sample: [[1]](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/storage/mod.rs#L565)  [[2]](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/storage/mod.rs#L581)   As you can see, the same ctxd (ContextDelegator) is shared for multiple futures, so that the compiler needs it to be `Sync`. However in fact these futures are never executed concurrently.

Reply:
Updated: Please check out this sample: [[1]](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/storage/mod.rs#L565)  [[2]](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/storage/mod.rs#L581) 

As you can see, the same `ctxd: ContextDelegators` is shared for multiple futures. Can we directly mark `ContextDelegators` as `Sync`? No. When we concurrently execute futures while in which `ContextDelegators` are shared, it is accessed **concurrently**. Notice, here we have concurrent access so that we'd better leave Rust to determine whether or not it is sync.

Then, let's take a look at what `ContextDelegators` wrapped, i.e. `ContextDelegator`. Is it sync? Yes! -- it is never accessed concurrently. The reason it is **never** accessed concurrently is because:
1. it is behind `HashMap<ThreadId>` and we use current thread id to get an instance, so that the instance we got is always belong to the current thread.
2. it is not public to the user (user can only access `Context` it wrapped, see pub methods) so that user does not have a chance to share it (neither has chance to send it as well).

According to the two points above, we can ensure that `ContextDelegator` does not have concurrent accesses and we can safely mark it as `Sync` to bypass the compiler check.

So finally, in fact, `ContextDelegators` is sync (due to `ContextDelegator` is sync), but I think it's better to mark `ContextDelegator` as `Sync`, because it is the root reason why `ContextDelegators` is sync.

Reply:
> ...ContextDelegators is sync (due to ContextDelegator is sync)...it is the root reason why ContextDelegators is sync...

That's wrong. A struct is `Sync` because it's safe when access it via the provided APIs concurrently which take a `&self`. `ContextDelegator` is not `Sync`, because when access it concurrently will lead to undefined result. You can't say it `Sync` because you know yourself will always use it in a safe way. You are not the only one user/maintainer of the struct you define. This is an abuse of the `Sync` marker.

Reply:
> You can't say it Sync because you know yourself will always use it in a safe way. You are not the only one user/maintainer of the struct you define.

As said before (see point 1 & point 2 above), the `ContextDelegator` is an internal structure of futurepool. Users does not have a chance to use it outside the `futurepool`. If you treat `futurepool` itself as a library and does not modify its code, when you use the `futurepool`, you simply cannot write out a piece of code to access `ContextDelegators` concurrently.

Reply:
And finally, I think marking a structure that will never concurrently accessed as `Sync` is correct, because if there is no concurrency according to the logic in reality, the `Sync` is simply nothing. It is invalid. There is no `Sync`.

Now `!Sync` makes the compiler rejects my code (but in fact these codes are not (and will never be) concurrently accessing), how would you do to make it compile without introducing overhead instead of marking it as `Sync`?

Reply:
> ...Any library can break if you modify its internal code without knowing how it works.

It's not the excuse to break the `Sync` check. The check is wrong no matter what the result is if prerequisite is wrong.

> ...that will never concurrently...

It's your design not your implementation, and that's the point of compile time check of rust, to detect such buggy code at compile time. I think the only thing you know right now is the struct
should not be accessed concurrently.

> ...Marking as Sync is fixing the compiler, telling it not to interfere this...

Compiler is correct about `ContextDelegator`. It's not `Sync`. But it's wrong about `ContextDelegators`, because the mechanism inside `ContextDelegators` makes it safe to be accessed concurrently via its API. So you should mark `ContextDelegators` as `Sync` instead of `ContextDelegator`.